apis/validator/beacon_committee_subscriptions.yaml

```diff
@@ -1,73 +1,36 @@
-create:
-  post:
-    operationId: "subscribeToBeaconCommitteeSubnet"
-    summary: Subscribe beacon node to commitee attestation subnet
-    description: "If validator is connected to multiple nodes, it should send this request to all of them to insure subnet network stability."
-    tags:
-      - MinimalValidatorApi
-      - Validator
-    requestBody:
-      content:
-        application/json:
-          schema:
-            type: array
-            items:
-              type: object
-              properties:
-                attestation_committee_index:
-                  $ref: '../../beacon-node-oapi.yaml#/components/schemas/Uint64'
-                slot:
-                  allOf:
-                    - $ref: '../../beacon-node-oapi.yaml#/components/schemas/Uint64'
-                    - description: "Should be slot at which validator is assigned to attest"
-                slot_signature:
-                  allOf:
-                    - $ref: '../../beacon-node-oapi.yaml#/components/schemas/Signature'
-                    - description: "Should be valid against given slot and validator pubkey."
-                aggregator_pubkey:
-                  $ref: '../../beacon-node-oapi.yaml#/components/schemas/PublicKey'
-    responses:
-      "200":
-        description: "Slot signature is valid and beacon node is subscribed to given committee attestation subnet."
-      "400":
-        $ref: '../../beacon-node-oapi.yaml#/components/responses/InvalidRequest'
-      "403":
-        description: "Validator is not aggregator"
-      "500":
-        $ref: '../../beacon-node-oapi.yaml#/components/responses/InternalError'
-      "503":
-        $ref: '../../beacon-node-oapi.yaml#/components/responses/CurrentlySyncing'
-
-attestations:
-  get:
-    summary: "Get unaggregated wire attestations"
-    operationId: "getWireAttestations"
-    description: "Endpoint should aggregate all collected attestations and produce AggregateAndProof without proof. Validator running on weak hardware is expected to add proof and publish"
-    tags:
-      - Validator
-    parameters:
-      - name: committee_index
-        in: path
-        required: true
-        description: Index of validator committee
+post:
+  operationId: "subscribeToBeaconCommitteeSubnet"
```

Comment:
From what I've seen everyone subscribes to random subnets by default (even though they don't have validators) so we don't need another endpoint for that. It's unclear how that should work, what would subsequent calls to that endpoint do? With few calls, you could subscribe BN to every subnet in which case BN would aggregate all attestations on all subnets which could be quite resource intensive.

Reply:
I agree with the idea of having external control of subscriptions to subnets, but if this was via an endpoint in a different namespace (_e.g._ debug) then wouldn't the fact that this subscription be for aggregation be inherent in the /validator namespace?

Or we could be less proscriptive, and rename it to something like `requiresBeaconCommitteeSubnet` to make it more of a notification and less of a command.

Reply:
Reason to allow manual subscription to subnets is that it gives a level of control in environments where the beacon node *may* be aggregating but isn't sure.

If you provided beacon nodes as a general service, for example, you'd want to subscribe to all subnets in case a client validator who happened to be an aggregator sent you an unaggregated attestation.

Reply:
> If you provided beacon nodes as a general service, for example, you'd want to subscribe to all subnets in case a client validator who happened to be an aggregator sent you an unaggregated attestation.

Depends on how much of a load on server and network does aggregating attestations on all subnets. If you are providing beacon nodes as a general service I imagine you would like to distribute that load and scale horizontally. [See my comment on load balancing](https://github.com/ethereum/eth2.0-APIs/pull/58#discussion_r449455046)

Even if you wan't to be subscribed on every subnet, I imagine config file or cli flag would be more appropriate to trigger that.

Reply:
The problem with config flags is that they are static, so hard to do dynamic load balancing strategies.

But this is somewhat off-topic for this particular item.  I do think that renaming this endpoint to be be a request or notification, rather than a command, fits better with its purpose and REST in general.

Reply:
> But this is somewhat off-topic for this particular item. I do think that renaming this endpoint to be be a request or notification, rather than a command, fits better with its purpose and REST in general.

So think endpoint it might be possible for BN to decline subscription to subnet?

Reply:
> So think endpoint it might be possible for BN to decline subscription to subnet?

I think that it's possible that the BN would be unable to fulfil the request.  If a user limited the BN's peers to a small number, for example, and none of its peers were subscribed to the requested subnet, the BN couldn't make a subscription.

However, because of the asynchronous nature of networking a request to subscribe would probably have to come back with either "I'm already subscribed" or "I'll try" rather than a definitive "subscribed" result.  Hence the idea of telling the beacon node what the validator requires.  And thinking about it, a companion API call along the lines of `isSubscribedToBeaconCommitteeSubnet()` may also be useful (because regardless of if the BN was subscribed to a subnet the the time the subscription was requested it doesn't mean it's still subscribed to it when the validator produces its attestation).

Reply:
> If a user limited the BN's peers to a small number, for example, and none of its peers were subscribed to the requested subnet, the BN couldn't make a subscription.

Whole point of this request is to allow BN to disconnect current peers and discover new one with requested subnet.

> isSubscribedToBeaconCommitteeSubnet

I think `hasBeaconCommitteeSubnet` peers would be more useful. Because you can subscribe to topic but don't have any peer broadcasting on that tobic, right?

> because regardless of if the BN was subscribed to a subnet the the time the subscription was requested it doesn't mean it's still subscribed to it when the validator produces its attestation

I don't think we should allow BN to behave in that way. After validator sends `subscribeToBeaconSubnet`, bn knows it's valid request from real validator, must find peers for that subnet (even if that means disconnecting current peers) and must be subscribed to that subnet until end of given slot. Only reason ofr BN to decline that request is if it's configured to run on certain sets of subnets.

Whats the reason for allowing BN to be picky about subnets and how long it's gonna be subscribed?

Reply:
> I don't think we should allow BN to behave in that way.

My general point is that the BN doesn't get the final say in the matter.  If it is limited in some external fashion (network provider takes exception to them, other peers are at max capacity, _etc._) then it may not be possible for the BN to either initiate or maintain a subscription.

And if we accept that this could be the case, then having calls for "I need you to subscribe to X" and "are you subscribed to X" are closer to what a validator client would need than assuming that a request to subscribe is honoured over any given time period.

Interesting point about if we should consider subscribed peers as subscribed: although it has the correct effect if the validator just cares about getting their attestation out, if they are also aggregating this wouldn't be enough.  We could have multiple calls (are you subscribed, can you forward to someone who is) but that distinction is quite subtle and may confuse implementers so I'd be wary of doing that.  I also have some (admittedly nebulous) worries about network fragility if BNs started to rely on other nodes (perhaps a single other node) being subscribed to a given subnet rather than subscribing themselves, which this could encourage.

Reply:
> then it may not be possible for the BN to either initiate or maintain a subscription.
 
This endpoint just forces BN:
- update enr with that subnet
- add subnet topic in gossipsub
- run a discv5 search for peers for that topic

Those stuff aren't externally affected, if he fails to find peers in given matter it doesn't mean he is not subscribed to subnet. In that case it will just return aggregator attestation or error if he doesn't have one.

We could add endpoint if BN has peers for that subnet so validator can try to ask some other BN for aggregated attestation